packages/components/src/toggle-group-control/toggle-group-control/as-button-group.tsx

```diff
@@ -46,39 +49,28 @@ function UnforwardedToggleGroupControlAsButtonGroup(
 		ToggleGroupControlAsButtonGroup,
 		'toggle-group-control-as-button-group'
 	).toString();
-	const [ selectedValue, setSelectedValue ] = useState( value );
-	const groupContext = {
-		baseId,
-		state: selectedValue,
-		setState: setSelectedValue,
-	};
 	const previousValue = usePrevious( value );
-
-	// Propagate groupContext.state change.
-	useUpdateEffect( () => {
-		// Avoid calling onChange if groupContext state changed
-		// from incoming value.
-		if ( previousValue !== groupContext.state ) {
-			onChange( groupContext.state );
-		}
-	}, [ groupContext.state ] );
-
-	// Sync incoming value with groupContext.state.
-	useUpdateEffect( () => {
-		if ( value !== groupContext.state ) {
-			groupContext.setState( value );
-		}
-	}, [ value ] );
-
+	const [ selectedValue, setSelectedValue ] = useControlledValue( {
+		defaultValue: previousValue,
+		onChange,
+		value,
+	} );
+	// Expose selectedValue getter/setter via context
+	const groupContext: ToggleGroupControlContextProps = useMemo(
+		() => ( {
+			baseId,
+			state: selectedValue,
+			setState: setSelectedValue as React.Dispatch<
+				React.SetStateAction< string | number | undefined >
+			>,
```

Comment:
Instead of typecasting `setSelectedValue`, we can try to type `useControlledValue` a bit most narrowly using [`const` assertions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions)

```diff
diff --git a/packages/components/src/toggle-group-control/toggle-group-control/as-button-group.tsx b/packages/components/src/toggle-group-control/toggle-group-control/as-button-group.tsx
index 0de5543bcc..a9e7879071 100644
--- a/packages/components/src/toggle-group-control/toggle-group-control/as-button-group.tsx
+++ b/packages/components/src/toggle-group-control/toggle-group-control/as-button-group.tsx
@@ -60,9 +60,7 @@ function UnforwardedToggleGroupControlAsButtonGroup(
 		() => ( {
 			baseId,
 			state: selectedValue,
-			setState: setSelectedValue as React.Dispatch<
-				React.SetStateAction< string | number | undefined >
-			>,
+			setState: setSelectedValue,
 			isBlock: ! isAdaptiveWidth,
 			isDeselectable: true,
 			size,
diff --git a/packages/components/src/utils/hooks/use-controlled-value.ts b/packages/components/src/utils/hooks/use-controlled-value.ts
index 5a3824d465..bb7f3aea80 100644
--- a/packages/components/src/utils/hooks/use-controlled-value.ts
+++ b/packages/components/src/utils/hooks/use-controlled-value.ts
@@ -22,7 +22,7 @@ export function useControlledValue< T >( {
 	defaultValue,
 	onChange,
 	value: valueProp,
-}: Props< T > ): [ T | undefined, ( value: T ) => void ] {
+}: Props< T > ) {
 	const hasValue = typeof valueProp !== 'undefined';
 	const initialValue = hasValue ? valueProp : defaultValue;
 	const [ state, setState ] = useState( initialValue );
@@ -40,5 +40,5 @@ export function useControlledValue< T >( {
 		setValue = setState;
 	}
 
-	return [ value, setValue ];
+	return [ value, setValue as typeof setState ] as const;
 }
```

Basically:
- in the return statement, we typecast `setValue` to be "whatever is the type of the `setState` function
- we add `as const`, which basically tells TypeScript to interpret the type of the return statement _as literally as possible_

Reply:
I can actually open a separate PR with this change, so that this PR stays focused on `ToggleGroupControl`

Reply:
https://github.com/WordPress/gutenberg/pull/46164

Reply:
Thank you for making that change. I'm updating the typecasting now.